python/lsst/verify/tasks/metadataMetricTask.py

```diff
@@ -56,23 +58,44 @@ class SingleMetadataMetricConnections(
         doc="The target top-level task's metadata. The name must be set to "
             "the metadata's butler type, such as 'processCcd_metadata'.",
         storageClass="PropertySet",
-        dimensions={"Instrument", "Exposure", "Detector"},
+        dimensions={"instrument", "visit", "detector"},
         multiple=False,
     )
 
+    def __init__(self, *, config=None):
+        """Customize the connections for a specific MetricTask instance.
+
+        Parameters
+        ----------
+        config : `MetadataMetricConfig`
+            A config for `MetadataMetricTask` or one of its subclasses.
+        """
+        super().__init__(config=config)
+        if config and config.metadataDimensions != self.metadata.dimensions:
+            # Hack, but only way to get a connection without fixed dimensions
+            newMetadata = connectionTypes.Input(
+                name=self.metadata.name,
+                doc=self.metadata.doc,
+                storageClass=self.metadata.storageClass,
+                dimensions=config.metadataDimensions,
+                multiple=self.metadata.multiple,
+            )
+            self.metadata = newMetadata
+            self.dimensions = config.metadataDimensions
```

Comment:
There is no way you would have gotten this w/o digging through the code, as this is not really the original intended behavior, but you should also add:
```python
self.allConnections['metadata'] = self.metadata
```
This is a dictionary used to track all the connections it is mostly unused now, but adding this will future proof things if it starts to be used in other areas.

Reply:
I cargo-culted this from [cp_pipe](https://github.com/lsst/cp_pipe/blob/32a200ec5c58a50283f7293ef49ecc410406115e/python/lsst/cp/pipe/cpCombine.py#L131), so I assume the same change needs to be made there?

Reply:
@natelust, when I add the new line, I get an error on running `ap_verify` that suggests the dimension override didn't take. I can't reproduce it in unit tests; the only difference I see is that `allConnections['metadata']` has a templated name `{labelName}_metadata` without the new line and has a substituted name with it.

I'm trying to trace back through the `ap_verify` run, but do you have any idea what might cause the behavior I'm seeing?

Reply:
I found part of the problem: apparently adding the line above causes/reveals some cross-talk between different task instances.

Specifically, if I have the following pipeline fragment:
```
    timing_imageDifference:
        class: lsst.verify.tasks.commonMetrics.TimingMetricTask
        config:
            connections.package: pipe_tasks
            connections.metric: ImageDifferenceTime
            connections.labelName: imageDifference
            metadataDimensions: [instrument, visit, detector, skymap]
            target: imageDifference.run
    timing_charImage:
        class: lsst.verify.tasks.commonMetrics.TimingMetricTask
        config:
            connections.package: pipe_tasks
            connections.metric: CharacterizeImageTime
            connections.labelName: charImage
            target: characterizeImage.run
    numUnassociatedDiaObjects:
        class: lsst.ap.association.metrics.NumberUnassociatedDiaObjectsMetricTask
        config:
            connections.labelName: diaPipe
```
then direct subtasks of `MetadataMetricTask` (e.g., [NumberUnassociatedDiaObjectsMetricTask](https://github.com/lsst/ap_association/blob/master/python/lsst/ap/association/metrics.py#L94)) is locked into an input named `imageDifference_metadata` (which then breaks because it doesn't have a dimensions override). The other instance of `TimingMetricTask` is unaffected:
```
(Pdb) for taskName, taskObject in byTask.items(): print(f"{taskName}: {taskObject.inputs}")
timing_imageDifference: {DatasetType(imageDifference_metadata, {abstract_filter, instrument, skymap, detector, physical_filter, visit_system, visit}, PropertySet)}
timing_charImage: {DatasetType(charImage_metadata, {abstract_filter, instrument, detector, physical_filter, visit_system, visit}, PropertySet)}
numUnassociatedDiaObjects: {DatasetType(imageDifference_metadata, {abstract_filter, instrument, detector, physical_filter, visit_system, visit}, PropertySet)}
```

Reply:
After some Slack discussion, fixed by adding a defensive copy of `allConnections` to lsst/pipe_base#142.